staging/src/k8s.io/api/storage/v1beta1/types.go

```diff
@@ -209,3 +209,203 @@ type VolumeError struct {
+}
+
+// VolumePluginMechanism is the type of mechanism components should use for volume operations
+type VolumePluginMechanism string
+
+const (
+	// VolumePluginMechanismInTree means components should use the in-tree plugin for volume operations
+	VolumePluginMechanismInTree VolumePluginMechanism = "InTree"
+	// VolumePluginMechanismCSI means components should use the CSI Driver for volume operations
+	VolumePluginMechanismCSI VolumePluginMechanism = "CSI"
+)
+
+// CSIDriverInfoStatus holds information about the status of one CSI driver installed on a node
+type CSIDriverInfoStatus struct {
+	// This is the name of the CSI driver that this object refers to.
+	// This MUST be the same name returned by the CSI GetPluginName() call for
+	// that driver. This field is required.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+
+	// available is a boolean representing whether the driver has been installed
+	// on this node or not. It is a way for kubelet to communicate to the
+	// kube-controller-manager and scheduler about the driver's availability.
+	// This field is required. When a driver is uninstalled, available is set to false.
+	Available bool `json:"available" protobuf:"varint,2,opt,name=available"`
+
+	// volumePluginMechanism announces what mechanism underlies volume plugins.
+	// It is set by Kubelet. It is used by the attach/detach controller, which
+	// needs to know how to perform attachments. The allowed values are:
+	// * "InTree": the volume operation (e.g., attach/detach) ought to be
+	//   directly performed by the attach/detach controller.
+	// * "CSI": the attach/detach controller ought to request
+	//   the csi plugin to perform the volume operation rather than perform it directly.
+	VolumePluginMechanism *VolumePluginMechanism `json:"volumePluginMechanism" protobuf:"bytes,3,opt,name=volumePluginMechanism,casttype=VolumePluginMechanism"`
+}
+
+// CSIDriverInfoSpec holds information about the specification of one CSI driver installed on a node
+type CSIDriverInfoSpec struct {
+	// This is the name of the CSI driver that this object refers to.
+	// This MUST be the same name returned by the CSI GetPluginName() call for
+	// that driver.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+
+	// nodeID of the node from the driver point of view.
+	// This field enables Kubernetes to communicate with storage systems that do
+	// not share the same nomenclature for nodes. For example, Kubernetes may
+	// refer to a given node as "node1", but the storage system may refer to
+	// the same node as "nodeA". When Kubernetes issues a command to the storage
+	// system to attach a volume to a specific node, it can use this field to
+	// refer to the node name using the ID that the storage system will
+	// understand, e.g. "nodeA" instead of "node1".
+	// This field must be populated. An empty string means NodeID is not initialized
```

Comment:
This reads a little weird.  Prefer simpler: "This field must be populated, and must not be empty".

Reply:
I think empty string is valid during migration.  I'll reword like this: This field must be populated. An empty string means NodeID is not initialized by the driver and it is a valid value.